### PR TITLE
Slice/Array intrinsics: `__slice` and `__elem_at`

### PR DESCRIPTION
**Pull request by xunilrj**
## Description

This PR is part of https://github.com/FuelLabs/sway/issues/5110 and introduces two new intrinsic: `__slice` and `__elem_at`.

`__slice` allows the creation of slices by slicing arrays or other slices. Whilst `__elem_at` returns a reference to an item inside the slice or the array.

## Out of bounds checks

These intrinsic will not generate any runtime checks, these must be done manually, when and where appropriate; but they do a complete static analysis of all indices, to avoid runtime buffer overflows, when possible.

That means that at runtime, it is possible to do a buffer overflow when reading/writing, which is an "undefined behaviour" as to what will happen.

## Empty Array

This PR also solves a problem with empty arrays. Before empty arrays such as `let a = []` were being type-checked as `[Never; 0]`, which means that any code after them was being marked as dead.

Now we correctly type check them as `[Unknown; 0]` and return a more friendly error.

```
          4 |
          5 |     // Empty array
          6 |     let a = [];
            |             ^^ Type must be known at this point
          7 | }
            |
          ____
```

## Check of constants inside fns

This PR also solves a problem with not checking `const` expressions inside `fns`. We, for example, do not allow slices in constants, but we were only checking globals. Now we check constants also inside functions, methods etc...

## Small improvements for our e2e

We can now `dbg` inside our e2e harness and get results like the ones below. One needs to include the lib `test/src/e2e_vm_tests/utils` and cal `something.dbg()` or `something.dbgln()`. There is no magic, and structs/enums will need to manually implement the `Dbg` trait.

This is only to facilitate the debugging of our e2e tests.

image

## Checklist

- [ ] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [ ] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by IGI-111 — COMMENTED** [review]
LGTM but these new intrinsics should be documented.

**Review by IGI-111 — DISMISSED** [review]

**Review by IGI-111 — APPROVED** [review]

File: test/src/e2e_vm_tests/mod.rs [hunk 1]
```diff
@@ -130,6 +133,28 @@ fn print_receipts(output: &mut String, receipts: &[Receipt]) {
                 is,
                 data,
             } => {
+                // Small hack to allow log from tests.
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply test/src/e2e_vm_tests/mod.rs:136]: Will do it in a next PR.

File: sway-ir/src/irtype.rs [hunk 2]
```diff
@@ -451,6 +461,15 @@ impl Type {
         }
     }
 
+    /// Get the type of the array element, if applicable.
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply sway-ir/src/irtype.rs:464]: Will do it in a next PR.

File: sway-core/src/lib.rs [hunk 3]
```diff
@@ -828,10 +830,13 @@ pub(crate) fn compile_ast_to_ir_to_asm(
...
-    // The IR pipeline relies on type information being fully resolved.
-    // If type information is found to still be generic or unresolved inside of
-    // IR, this is considered an internal compiler error. To resolve this situation,
-    // we need to explicitly ensure all types are resolved before going into IR.
+    // IR generaterion requires type information to be fully resolved.
+    //
+    // If type information is found to still be generic inside of
+    // IR, this is considered an internal compiler error.
+    //
+    // But, there are genuine cases for types be unknown here, like `let a = []`. These should
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply sway-core/src/lib.rs:838]: Will do it in a next PR.

File: docs/book/src/reference/compiler_intrinsics.md [hunk 4]
```diff
@@ -325,3 +331,44 @@ __jmp_mem()
...
+```sway
+__elem_at(item: &[T; N], index: u64) -> &T
+__elem_at(item: &[T], index: u64) -> &T
+__elem_at(item: &mut [T; N], index: u64) -> &mut T
+__elem_at(item: &mut [T], index: u64) -> &mut T
+```
+
+**Description:** Returns a reference to the indexed element. The mutability of reference is defined by the first parameter mutability.
+
+Runtime bound checks are not generated, and must be done manually when and where appropriated. Compile time bound checks are done when possible.
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply docs/book/src/reference/compiler_intrinsics.md:369]: Will do it in a next PR.

File: docs/book/src/reference/compiler_intrinsics.md [hunk 5]
```diff
@@ -325,3 +331,44 @@ __jmp_mem()
...
+__slice(item: &mut [T; N], start: u64, end: u64) -> &mut [T]
+__slice(item: &mut [T], start: u64, end: u64) -> &mut [T]
+```
+
+**Description:** Slices an array or another slice.
+
+This intrinsic returns a reference to a slice containing the range of elements inside `item`.
+The mutability of reference is defined by the first parameter mutability.
+
+Runtime bound checks are not generated, and must be done manually when and where appropriated. Compile time bound checks are done when possible.
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply docs/book/src/reference/compiler_intrinsics.md:349]: Will do it in a next PR.

**Review by kayagokalp — APPROVED** [review]

**Pull request merged**